libmscore/changeMap.cpp

```diff
@@ -0,0 +1,487 @@
+                              }
+                        }
+
+                  lastFix = event.tick;
+                  }
+            }
+
+      //
+      // Stage 2: readjust lengths of any colliding ramps:
+      //
+      int j = -1;
+      for (auto& event : values()) {
+            if (event.type != ChangeEventType::RAMP)
+                  continue;
+
+            j++;
+            if (!startsInRamp[j])
+                  continue;
+
+            // Take a copy of the event...
+            auto tempChangeEvent = event;
+            remove(event.tick, event);
+            tempChangeEvent.tick = endPoints[j-1].second;
+            // ...and move it to its new tick position
+            insert(tempChangeEvent.tick, tempChangeEvent);
+            }
+
+      //
+      // Stage 3: cache start and end values for each ramp
+      //
+      for (auto i = begin(); i != end(); i++) {
+            auto& event = i.value();
+            if (event.type != ChangeEventType::RAMP)
+                  continue;
+
+            // Phase 1: cache a start value for the ramp
+            // Try and get a fix at the tick of this ramp
+            bool foundFix = false;
+            for (auto& currentChangeEvent : values(event.tick)) {
+                  if (currentChangeEvent.type == ChangeEventType::FIX) {
+                        event.cachedStart = currentChangeEvent.value;
+                        foundFix = true;
+                        break;
+                        }
+                  }
+
+            // If there isn't a fix, use from the last event:
+            //  - the cached end value if it's a ramp
+            //  - the value if it's a fix
+            if (!foundFix) {
+                  if (i != begin()) {
+                        auto prevChangeEventIter = i;
+                        prevChangeEventIter--;
+
+                        // Look for a ramp first
+                        bool foundRamp = false;
+                        for (auto& prevChangeEvent : values(prevChangeEventIter.key())) {
+                              if (prevChangeEvent.type == ChangeEventType::RAMP) {
+                                    event.cachedStart = prevChangeEvent.cachedEnd;
+                                    foundRamp = true;
+                                    break;
+                                    }
+                              }
+
+                        if (!foundRamp) {
+                              // prevChangeEventIter must point to a fix in this case
+                              event.cachedStart = prevChangeEventIter.value().value;
+                              }
+                        }
+                  else {
+                        event.cachedStart = DEFAULT_VALUE;
+                        }
+                  }
+
+            // Phase 2: cache an end value for the ramp
+            // If there's no set velocity change:
+            if (event.value == 0) {
+                  auto nextChangeEventIter = upperBound(event.tick);
```

Comment:
Why not `auto nextEventIter = i; ++nextEventIter`, like it is done for `prevEventIter`? `upperBound` will search this event from scratch.

Also, should this event be "next" to `tick` or `etick`? If we are going to fill the value at event's end it looks like the answer should be `etick`, but `tick` is used here.

Reply:
If etick were used with upperBound, then we would miss a fix exactly at the end of the ramp (since if upperBound matches the value exactly, it will select the next value on). You're right though about just taking a copy of `i` and incrementing it, that would do the same thing more efficiently.